### DISCUSSION TITLE
Does PyMuPDF support TOC creation from a pdf document?

### DISCUSSION THREAD

**Discussion by tikitong** [pymupdf/PyMuPDF#2015] [Q&A] [answered]
First of all thank you very much for this great work. I particularly appreciate your layout preserving text extraction method. 
My question is: does pymupdf support TOC creation for a pdf document? 

`get_toc()` method used such as:

```python
import fitz

pdf_filename = 'my.pdf'
with fitz.open(pdf_filename) as doc:
    print(doc.get_toc())
```
seems to give results only if the TOC is already present at the beginning of the document.

On this pdf: https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/Dart.pdf for example, is there a method in pymupdf to generate the pdf outline? In this case only **the numbered titles of the paragraphs**. 

As well as for pdf with more complex formatting such as: https://blog.xpgreat.com/file/lstm.pdf, with numbered parts and sub-parts. 

I also tested [mupdf](https://mupdf.com/docs/manual-mutool-show.html) with the command: `mutool show my.pdf outline` but it returns nothing with or without TOC inside the pdf file in my case. 

## Your configuration (mandatory)
In my case, I made installation on macOS **arm64 M2** (not Intel). 
I create a conda osx-64 environment inspired from [this](https://stackoverflow.com/a/70219965/14864907) amazing solution, and its work for me. 
```
conda create -n pymupdf
conda activate pymupdf
conda config --env --set subdir osx-64
conda install python=3.9 
python -m pip install --upgrade pymupdf
```
`print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` gives me:
```
3.9.13 (main, Oct 13 2022, 16:12:30) 
[Clang 12.0.0 ] 
 darwin 
 
PyMuPDF 1.20.2: Python bindings for the MuPDF 1.20.3 library.
Version date: 2022-08-13 00:00:01.
Built for Python 3.9 on darwin (64-bit).
```
Please feel free to modify the [README.md](https://github.com/pymupdf/PyMuPDF#installation)  to notify macOS users with the apple chip that it also works by following this steps, I'm sure it will be useful for some :).

**Comment by JorjMcKie** [comment]
This is a typical "Discussions" item, so let me convert it before asnwering.

**Comment by JorjMcKie** [comment, marked as answer]
Talking about TOC in PDF can easily become confusing:
1. A PDF may have a special internal structure of interrelated objects, with forward, backward, downward, upward pointers. PDF viewers show this structure on the left in a separate detail windows, where you can click on single items for fast access of the respective page.
2. Then there may also (or only) exist pages in the PDF's content, showing a "printed" version of that structure. In most cases, the single lines in this area are PDF internal links. So clicking on them will also jump to the respective page.

A PDF technically can have both of the above things, or just one, or neither. If neither exists or just number 2 above, then `doc.get_toc()` will be an empty list - of course.

I think you are asking for a way to create number 1 if just number 2 exists, right?

Because TOC **_pages_** in a PDF are just ordinary text, it is left to the programmer's wits to
1. identify those pages (which may not be trivial at all in itself!), and
2. programmatically turn each line item on a TOC page into an item acceptable for `doc.set_toc()`. This too may be quite challenging.

And there certainly also is no function in PyMuPDF which creates a TOC if nothing of the above exists at all.
I think you were asking for something that interprets things like section headers, text properties (bold) and similar.
No this is left to you as a programmer. I think I saw your similar question on StackOverflow the other day.

-------
> seems to give results only if the TOC is already present ~at the beginning of the document.~

Correct ... with the redaction.

**Reply by fabiorzfreitas** [reply]
> I think you are asking for a way to create number 1 if just number 2 exists, right?

Hi, I've been looking for something similar for some time now. I'm trying to do the opposite, create a TOC from my Outlines, which are properly ordered and numbered, with no nesting (although none of that matters, right?).

The reason I'm trying to achieve this is that some mobile PDF readers do not show Outlines at all, but I happen to have a document in which indexing is very much needed.

This probably looks like a stupid question, since `Document.get_toc()` is very clearly available, but I'm not knowledgeable enough to implement it in a project of my own. I think perhaps I actually could do so if I had lots of hours to research, but there's probably either already a complete tool for this, or a very trivial way for someone more acquainted with PyMuPDF to do it!

Thanks :)

**Reply by JorjMcKie** [reply]
@fabiorzfreitas - you simply provide a list of lists like so:
```python
toc = [
    [lvl, title, page-1-based],  # page number must be 1-based
    [lvl, title, page-1-based],
    [lvl, title, page-1-based],
    ...
]
```
Then `doc.set_toc(toc)`. That's all.
There are a few rules for the values of lvl (level) - please see the documentation for details.

**Reply by JorjMcKie** [reply]
"Nesting" is nothing but indentation - which is controlled by the lvl value.

**Reply by wabiloo** [reply]
@JorjMcKie 

> I think you are asking for a way to create number 1 if just number 2 exists, right?

But what if number 2 is not just a printed version, but actually contains working hyperlinks (because the PDF doc was generated from an HTML file)?
My PDF reader seems to know what page the hyperlinks go to...
![Screenshot 2024-04-05 at 12 51 00](https://github.com/pymupdf/PyMuPDF/assets/4056263/b9bb550f-6b84-4cb3-b2b7-e77b086ab4dc)

Is there a way of going through those hyperlinks, and finding the page number that they target, and on that basis generate the `toc` list of lists?

**Reply by JorjMcKie** [reply]
Hyperlinks can be extracted via `page.get_links()`. Gives a list of dictionaries which has a "page" key ...

**Reply by wabiloo** [reply]
Yep, I'm currently experimenting with it. Not super-easy, but my document has the advantage of using `LINK_NAMED`

**Reply by bsdcpp** [reply]
Is there a method in pymupdf to generate a printable toc page like number 2?
Just like the one starting on the fourth page of "mupdf_explored.pdf", 
it’s very beautiful and seems to be made using Latex.
Or I need to write my own toc pages with title, page numbers and links.
Thank you.

**Reply by JorjMcKie** [reply]
@bsdcpp You should have a look at the features of the [Story](https://pymupdf.readthedocs.io/en/latest/story-class.html#Story) class.
Your intention looks like a typical use case for it. You would have to prepare the formatting of the TOC line items of course. But outputting them on predefined rectangles is a thing that Story has been built for.
Once the pages produced by such an algorithm are ready, prepend (or append) them the document in question.

**Reply by bsdcpp** [reply]
Thanks a lot for the quick reply.
Before, I've been using qpdf/gs for decryption(Some files have permissions set), pdfrw for processing, and also cooperated with pdfkit to process some HTML. 
Now I can put them together into pymupdf, which is much more efficient.

**Reply by JorjMcKie** [reply]
Glad you like it!

**Reply by bsdcpp** [reply]
After reading the documentation, I found that there's already a good sample at    [docs/samples/code-printer.py](https://github.com/pymupdf/PyMuPDF/blob/main/docs/samples/code-printer.py), which is really great.

**Comment by tikitong** [comment]
Thank you for this explanation, I had difficulty in describing the problem precisely. 
>I think you were asking for something that interprets things like section headers, text properties (bold) and similar. 

It is exactly that! 

>No this is left to you as a programmer.

Yes, that sounds exciting! Before starting a development, I try to find out what has already been done to solve this problem. 

1. Is this an issue that has been raised before? I'm sure I'm not the first person to try to find a solution to this task.  

2. Thus, do you know if there are already solutions that come close to this goal or any libraries?  
 
3. I think that the [pdf.tocgen](https://github.com/Krasjet/pdf.tocgen) solution is able to generate the TOC **without** it being present  in the PDF, and apparently without internal links too, as show with [its example](https://krasjet.com/voice/pdf.tocgen/#a-worked-example). It is however specified that it extracts metadata (font attributes, positions), it is not pointers in the pdf, right? When I open the pdf from my side, it doesn't seem to meet either of your two above conditions. Have you already used this solution?

**Reply by JorjMcKie** [reply]
1. No. There have been people occasionally, who thought that things like a header or a footer must be miraculously identifiable in a PDF - which is not case of course, although that feature is contained in the official PDF specs, just nobody is ever using it.
2. and 3.: No, and I do not know that solution or package either. Looking into it shows that is using PyMuPDF ... and confirming that finding a solution to your problem is by no means trivial.

**Reply by tikitong** [reply]
Well noted, thanks again for the information and details. 

Edit : @JorjMcKie I find this question very interesting. Have you ever thought about undertaking the development of a solution for this task with the PyMuPDF team?

**Comment by JorjMcKie** [comment]
I once (long ago) have written a [GUI script](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/PDFoutline.py) which interactively allows to create a TOC:
It displays the pages on the right windows and offers to add / edit the lines of a TOC on the left windows.
May you can use it. But be prepared that there may be recent API changes that this scripts has no knowledge of.

**Reply by tikitong** [reply]
Yes I had seen this script while looking in PyMuPDF, I will think about developing a solution.
